// @shootismoke/ui/lib/util/pollutant.js
"use strict";
// Sh**t! I Smoke
// Copyright (C) 2018-2021  Marcelo S. Coelho, Amaury M.
exports.__esModule = true;
exports.primaryPollutant = exports.getAQI = exports.getPollutantData = void 0;
// Sh**t! I Smoke is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// Sh**t! I Smoke is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with Sh**t! I Smoke.  If not, see <http://www.gnu.org/licenses/>.
var convert_1 = require("@shootismoke/convert");
/**
 * Definitions and effects of various pollutants.
 *
 * @see https://www.sciencedirect.com/topics/chemistry/air-pollutant
 */
var pollutantData = {
    bc: {
        effects: 'Black carbon is a potent climate-warming component of particulate matter formed by the incomplete combustion of fossil fuels, wood and other fuels.',
        name: 'Black Carbon'
    },
    c6h6: {
        effects: 'Hydrocarbons are the primary pollutants that produce photochemical smog.',
        name: 'Hydrocarbons'
    },
    ch4: {
        effects: 'Partly responsible for the atmospheric greenhouse effect.',
        name: 'Methane'
    },
    co: {
        effects: 'Carbon monoxide reduces the oxygen-carrying capacity of the blood by combining with haemoglobin, thus deprives tissues of O2.',
        name: 'Carbon monoxide'
    },
    // FIXME Add co2
    // co2: {
    // 	effects: 'Partly responsible for the atmospheric greenhouse effect',
    // },
    no: {
        effects: 'Nitrogen oxides cause eye, throat, and lung irritation. Primary pollutants that produce photochemical smog and acid rain, destroy ozone at the stratosphere.',
        name: 'Nitrogen oxides'
    },
    no2: {
        effects: 'Nitrogen oxides cause eye, throat, and lung irritation. Primary pollutants that produce photochemical smog and acid rain, destroy ozone at the stratosphere.',
        name: 'Nitrogen oxides'
    },
    o3: {
        effects: 'Ozone causes eye, throat, and lung irritation, impairs lung function.',
        name: 'Ozone'
    },
    pm10: {
        effects: 'Particulate matters under 10μm (PM10), may cause breathing difficulties.',
        name: 'Particulates'
    },
    pm25: {
        effects: 'Particulate matters under 2.5μm (PM2.5), may cause breathing difficulties.',
        name: 'Particulates'
    },
    so2: {
        effects: 'Sulfur dioxide causes eye, throat, and lung irritation. Primary pollutants that produce acid rain.',
        name: 'Sulfur dioxide'
    }
};
/**
 * Get metadata about a pollutant.
 *
 * @param pollutant - The pollutant to get the data.
 */
function getPollutantData(pollutant) {
    var polData = pollutantData[pollutant];
    if (!polData) {
        throw new Error('All pollutants are in pollutantData. qed.');
    }
    return polData;
}
exports.getPollutantData = getPollutantData;
/**
 * From a set of OpenAQResults pollutant data, filter only the ones that can be
 * converted to USA EPA, and sort the set.
 *
 * @param results - The OpenAQResults data for all pollutants.
 */
function getSortedOpenAQResults(results) {
    // We attempt to sort the pollutants by AQI.
    var unsorted = results
        // Since we can only convert ugm3, we filter those results.
        .filter(function (_a) {
        var unit = _a.unit;
        return unit === convert_1.ugm3;
    })
        // Only these pollutants can be converted to usaEpa
        .filter(function (_a) {
        var parameter = _a.parameter;
        return ['o3', 'pm10', 'pm25', 'co', 'so2', 'no2'].includes(parameter);
    });
    // Sort the array by AQI.
    unsorted.sort(function (a, b) {
        return (0, convert_1.convert)(b.parameter, b.unit, 'usaEpa', b.value) -
            (0, convert_1.convert)(a.parameter, a.unit, 'usaEpa', a.value);
    });
    return unsorted;
}
/**
 * From a set of OpenAQResults pollutant data, get the AQI, that is, the AQI of
 * the primary pollutant.
 *
 * @param OpenAQResults - The OpenAQResults data for all pollutants.
 */
function getAQI(results) {
    var sorted = getSortedOpenAQResults(results);
    if (sorted[0]) {
        try {
            return (0, convert_1.convert)(sorted[0].parameter, sorted[0].unit, 'usaEpa', sorted[0].value);
        }
        catch (err) {
            // convert() throws if the unit is not supported. We just silently
            // return an error.
            return;
        }
    }
}
exports.getAQI = getAQI;
/**
 * From a set of OpenAQResults pollutant data, find the primary pollutant.
 *
 * @param results - The OpenAQResults data for all pollutants.
 */
function primaryPollutant(results) {
    var sorted = getSortedOpenAQResults(results);
    if (sorted[0]) {
        return sorted[0];
    }
}
exports.primaryPollutant = primaryPollutant;
