// @shootismoke/ui/lib/util/theme.js
"use strict";
// Sh**t! I Smoke
// Copyright (C) 2018-2021  Marcelo S. Coelho, Amaury M.
exports.__esModule = true;
exports.withPadding = exports.withLetterSpacing = exports.title = exports.text = exports.shitText = exports.link = exports.elevationShadowStyle = exports.disabledOpacity = exports.spacing = exports.secondaryTextColor = exports.textColor = exports.primaryColor = exports.iconBackgroundColor = exports.backgroundColor = void 0;
exports.backgroundColor = '#FAFAFC';
exports.iconBackgroundColor = '#EBE7DD';
exports.primaryColor = '#F8A65D';
exports.textColor = '#44464A';
exports.secondaryTextColor = '#8B909A';
exports.spacing = {
    tiny: 5,
    mini: 10,
    small: 15,
    normal: 20,
    big: 36
};
/**
 * Opacity for views that are disabled.
 */
exports.disabledOpacity = 0.3;
/**
 * Get consistent shadows between iOS and Android
 * @see https://stenbeck.io/styling-shadows-in-react-native-ios-and-android/
 */
function elevationShadowStyle(elevation, position) {
    if (position === void 0) { position = 'bottom'; }
    return {
        elevation: elevation,
        shadowColor: 'black',
        shadowOffset: {
            width: 0,
            height: (position === 'bottom' ? 1 : -1) * elevation
        },
        shadowOpacity: exports.disabledOpacity,
        shadowRadius: 0.8 * elevation
    };
}
exports.elevationShadowStyle = elevationShadowStyle;
exports.link = {
    color: exports.primaryColor,
    textDecorationLine: 'underline'
};
/**
 * Big text with "Sh*t! I smoked...""
 */
exports.shitText = {
    color: exports.textColor,
    fontSize: 31,
    fontWeight: '800',
    letterSpacing: -1,
    lineHeight: 36
};
/**
 * Normal text
 */
exports.text = {
    color: exports.secondaryTextColor,
    fontSize: 11,
    letterSpacing: 0.85,
    lineHeight: 15
};
exports.title = {
    letterSpacing: 3.14,
    lineHeight: 18,
    color: exports.textColor,
    fontSize: 12,
    fontWeight: '800'
};
exports.withLetterSpacing = {
    letterSpacing: 2
};
exports.withPadding = {
    paddingHorizontal: exports.spacing.normal
};
